includes/class-wc-form-handler.php

```diff
@@ -336,7 +336,13 @@ public static function checkout_action() {
 
 			wc_maybe_define_constant( 'WOOCOMMERCE_CHECKOUT', true );
 
-			WC()->checkout()->process_checkout();
+			// If cart has been modified since the user arrived to the checkout page and
+			// the page has not been updated, print notice for the user and don't process checkout.
+			if ( $_COOKIE['woocommerce_cart_dirty'] !== '0' && empty( $_POST['woocommerce_checkout_update_totals'] ) ) {
```

Comment:
I think `! empty` on the cart dirty cookie might be clearer and will avoid issues with int/string `0`.